core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -99,101 +88,122 @@ public CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionD
-                                                    }
-                                                    result.put(topicIdPartition, partitionData);
-                                                });
-                                    } finally {
-                                         sharePartition.releaseFetchLock();
-                                    }
-                                }
-                            });
-                            shareFetchPartitionData.future.complete(result);
-                            // we are releasing the lock to move ahead with the next request in queue.
-                            releaseProcessFetchQueueLock();
-                            if (!fetchQueue.isEmpty())
-                                maybeProcessFetchQueue();
-                            return BoxedUnit.UNIT;
-                        });
-            } finally {
-                if (isProcessFetchQueueLockAvailable())
-                    releaseProcessFetchQueueLock();
+        if (!processFetchQueueLock.compareAndSet(false, true)) {
+            // The queue is already being processed hence avoid re-triggering.
+            return;
+        }
+
+        // Initialize the topic partitions for which the fetch should be attempted.
+        Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = new LinkedHashMap<>();
+        ShareFetchPartitionData shareFetchPartitionData = fetchQueue.poll();
+        try {
+          assert shareFetchPartitionData != null;
+          shareFetchPartitionData.topicIdPartitions.forEach(topicIdPartition -> {
+                // TODO: Fetch inflight and delivery count from config.
+                SharePartition sharePartition = partitionCacheMap.computeIfAbsent(sharePartitionKey(
+                    shareFetchPartitionData.groupId, topicIdPartition),
+                    k -> new SharePartition(shareFetchPartitionData.groupId, topicIdPartition, 100, 5));
+                int partitionMaxBytes = shareFetchPartitionData.partitionMaxBytes.getOrDefault(topicIdPartition, 0);
+                // Add the share partition to the list of partitions to be fetched only if we can
+                // acquire the fetch lock on it.
+                if (sharePartition.maybeAcquireFetchLock()) {
+                    topicPartitionData.put(topicIdPartition, new FetchRequest.PartitionData(
+                        topicIdPartition.topicId(),
+                        sharePartition.nextFetchOffset(),
+                        -1,
+                        partitionMaxBytes,
+                        Optional.empty()));
+                }
+            });
+
+            if (topicPartitionData.isEmpty()) {
+                // No locks for share partitions could be acquired, so we complete the request and
+                // will re-fetch for the client in next poll.
+                shareFetchPartitionData.future.complete(Collections.emptyMap());
+                return;
```

Comment:
shouldn't we also call  `releaseProcessFetchQueueLock()` method to release fetch queue lock here?

Reply:
I didn't add the check as anyways some request should be in-flight. But probably shall do that, we might need to do better here later.

Reply:
Added.